File: cpp/0784.0_Letter_Case_Permutation.cpp
/*
执行用时：0 ms, 在所有 C++ 提交中击败了100.00% 的用户
内存消耗：9.9 MB, 在所有 C++ 提交中击败了75.66% 的用户
通过测试用例：63 / 63
*/
class Solution {
public:
    vector<string> letterCasePermutation(string s) {
        vector<string> ans;
        queue<string> q;
        q.emplace("");
        while (!q.empty()) {
            string &curr = q.front();
            if (curr.size() == s.size()) {
                ans.emplace_back(curr);
                q.pop();
            } else {
                int pos = curr.size();
                if (isalpha(s[pos])) {
                    string next = curr;
                    next.push_back(s[pos] ^ 32);
                    q.emplace(next);
                }
                curr.push_back(s[pos]);
            }
        }
        return ans;
    }
};

File: csharp/0784.0_Letter_Case_Permutation.cs
/*
执行用时：160 ms, 在所有 C# 提交中击败了26.19% 的用户
内存消耗：64.3 MB, 在所有 C# 提交中击败了73.81% 的用户
通过测试用例：63 / 63
*/
public class Solution {
    public IList<string> LetterCasePermutation(string s) {
        IList<string> ans = new List<string>();
        Queue<StringBuilder> q = new Queue<StringBuilder>();
        q.Enqueue(new StringBuilder());
        while (q.Count > 0) {
            StringBuilder curr = q.Peek();
            if (curr.Length == s.Length) {
                ans.Add(curr.ToString());
                q.Dequeue();
            } else {
                int pos = curr.Length;
                if (char.IsLetter(s[pos])) {
                    StringBuilder next = new StringBuilder(curr.ToString());
                    next.Append((char) (s[pos] ^ 32));
                    q.Enqueue(next);
                }
                curr.Append(s[pos]);
            }
        }
        return ans;
    }
}
